elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/HttpExecutable.scala

```diff
@@ -0,0 +1,49 @@
+package com.sksamuel.elastic4s.http
+
+import com.sksamuel.exts.Logging
+import org.apache.http.HttpEntity
+import org.elasticsearch.client.{Response, ResponseListener, RestClient}
+
+import scala.concurrent.{Future, Promise}
+
+/**
+  * @tparam T the type of the request object handled by this handler
+  * @tparam U the type of the response object returned by this handler
+  */
+trait HttpExecutable[T, U] extends Logging {
+
+  def execute(client: RestClient, request: T): Future[U]
+
+  // convenience methods to allow implementations of HttpExec2 to execute
+  // a HTTP request and receive a Scala Future.
+  implicit class RichRestClient(client: RestClient) {
+
+    import scala.collection.JavaConverters._
+
+    private def future(callback: ResponseListener => Unit, handler: ResponseHandler[U]): Future[U] = {
+      val p = Promise[U]()
+      callback(new ResponseListener {
+        override def onSuccess(r: Response): Unit = p.tryComplete(handler.onResponse(r))
+        override def onFailure(e: Exception): Unit = p.tryComplete(handler.onError(e))
+      })
+      p.future
+    }
+
+    def future(method: String,
```

Comment:
Why did you rename this from `performAsyncResponse` to `future`? This method name is a bit confusing. Some API docs would  be nice on this one.